### PR TITLE
dns/ddclient - add "get" protocol in custom service type

### PR DESCRIPTION
**Pull request by DaCookie4u**
I am loving this new "custom POST" option, but my provider unfortunately only supports GET-requests. POST-requests fail with a CSRF validation.

Since I am unable to change the code on my providers side I decided to implemented a "custom GET" option. This does the same work only with a GET-request instead of POST.

I tried to keep the changes as small as possible. Not sure if that is wanted, but this way 99% of the current code is used and only the request-type changes.

### CODE REVIEW

File: dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml [hunk 1]
```diff
@@ -51,8 +51,7 @@
                         <dinahosting>dinahosting</dinahosting>
                         <dnsmadeeasy>DNS Made Easy (digicert)</dnsmadeeasy>
                         <dns-o-matic>DNS-O-Matic</dns-o-matic>
-                        <dnsexit2>DNSExit API</dnsexit2>
-                        <dnsexit>DNSExit Legacy</dnsexit>
+                        <dnsexit>DNSExit</dnsexit>
```
Comment (fichtner) [inline dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml:54]: you should either rebase or fix with a merge from master

Comment (DaCookie4u) [reply dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml:54]: Yeah, sorry for the mess ... I always mess git up as soon as I try to roll something back :) I hope it's OK though ...

Comment (fichtner) [reply dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml]: no worries, I fixed it manually

**Comment by fichtner** [comment]
@DaCookie4u thanks, at first glance this looks good. @AdSchellevis will let you know about cleanups in a bit. He also mentioned we could add PUT as well while at it.


Cheers,
Franco

File: dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml [hunk 2]
```diff
@@ -96,6 +96,7 @@
                         <dyndns1>DynDns1</dyndns1>
                         <dyndns2>DynDns2</dyndns2>
                         <post>custom POST</post>
+                        <get>custom GET</get>
```
Comment (fichtner) [inline dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml:99]: sort and captialize

```suggestion
                        <get>Custom GET</get>
                        <post>Custom POST</post>
```

File: dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.php [hunk 3]
```diff
@@ -56,7 +56,7 @@ public function performValidation($validateFullModel = false)
                 continue;
             }
             $srv = (string)$node->server;
-            if  ((string)$node->protocol == 'post') {
+            if  (in_array((string)$node->protocol, array('post', 'get'))) {
```
Comment (fichtner) [inline dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.php:59]:
```suggestion
            if  (in_array((string)$node->protocol, ['get', 'post']))) {
```

File: dns/ddclient/src/opnsense/scripts/ddclient/lib/account/dyndns2.py [hunk 4]
```diff
@@ -65,11 +65,12 @@ def match(account):
     def execute(self):
         if super().execute():
             protocol = self.settings.get('protocol', None)
-            if protocol == 'post':
+            if protocol in [ 'post', 'get' ]:
```
Comment (fichtner) [inline dns/ddclient/src/opnsense/scripts/ddclient/lib/account/dyndns2.py:68]:
```suggestion
            if protocol in [ 'get', 'post' ]:
```

File: dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.php [hunk 5]
```diff
@@ -56,7 +56,7 @@ public function performValidation($validateFullModel = false)
                 continue;
             }
             $srv = (string)$node->server;
-            if (in_array((string)$node->protocol, ['get', 'post']))) {
+            if  (in_array((string)$node->protocol, ['get', 'post', 'put']))) {
```
Comment (fichtner) [inline dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.php:59]:
```suggestion
            if (in_array((string)$node->protocol, ['get', 'post', 'put']))) {
```

Comment (fichtner) [reply dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.php:59]: now the whitespace is back ;)

**Review by fichtner — APPROVED** [review]
thanks, LGTM

**Comment by DaCookie4u** [comment]
I liked the PUT idea, so I put it in. Hope that was OK.

I also switched the getattr() method to the general [request](https://requests.readthedocs.io/en/latest/api/#requests.request) method. I have the feeling that is cleaner and somehow safer.

**Comment by DaCookie4u** [comment]
BTW, how long does it for plugins to be packaged to the main repository after being committed to master? I couldn't find any info on the exact process that leads from master to package.

**Comment by fichtner** [comment]
master branch merges land in the next development version of the upcoming release. If we can wrap up tomorrow that would be 23.7.2 for early next week.

But I think these changes are simple enough to be pushed to stable as well so they will be released directly in 23.7.2.

**Comment by AdSchellevis** [comment]
@DaCookie4u looks good, thanks!

**Pull request merged**